### ISSUE TITLE
Lifting logic

### ISSUE THREAD

**Issue by davidagold** [davidagold/_AbstractTables.jl#2]
This issue concerns the implementation of the `collect` machinery for types `T <: AbstractTable` that satisfy the query/column-indexable interfaces. When collecting, say, a `q::SelectNode` over a data source of such a type `T`, the `collect` machinery will apply the lambda stored in `q.helper::SelectHelper` over the (`zip`ed) relevant columns of the data source. More concretely, suppose `q` represents the query

``` julia
@query select(tbl, C = A * B)
```

Then the aforementioned lambda is of the form `tuple -> tuple[1] * tuple[2]` and is applied to each element of `zip(tbl[:A], tbl[:B])`. Note that `tbl[:A]`, `tbl[:B]` are assumed to be `NullableArray` columns, and so the application of the lambda involves lifting the latter over elements of this zipped iterator.

Currently, lifting is achieved in a "wholesale" fashion: the entire AST of the row-wise kernel (in this case, `A * B`) is encoded without modification (modulo processing interpolated quantities) in the aforementioned lambda, which is then lifted as follows: the `collect` machinery checks to see if any of the elements of the tuple returned by the zipped iterator is null, and if so the machinery does not apply the lambda but rather inserts a null value into the resultant column. If all of the elements of the tuple are non-null, then the machinery applies the lambda to the "unwrapped" tuple, i.e. the tuple of values stored inside the non-null `Nullable`s. Thus the present lifting regime has the "standard" lifting semantics hard-coded.

However, this may not be appropriate in all cases. In particular, and apart from functions that are intended to be applied to the `Nullable` container itself (e.g.., `isnull`), users may desire the three-valued logic (3VL) semantics in their queries. This requires a more flexible and more extensive lifting regime.

The strategy that John has proposed involves modifying the AST of transformations appearing in the arguments to `@query` verbs, replacing each function call `f(xs...)` with `lift(f, xs...)`, where `lift` can encode the standard lifting semantics for general `f`s and specialized lifting semantics for functions like `isnull` and the logical operators. 

The above will allow us to propagate 3VL invariants with certain limitations. Since we can only modify ASTs that are visible to the `@query` macro, we can't extend 3VL semantics to logical operators called inside of functions appearing in transformations in `@query` manipulations. For instance, the above strategy would allow us give 3VL semantics to

``` julia
@query select(tbl, A | B)
```

but would employ standard lifting semantics in the collection of 

``` julia
f(x, y) = x | y
@query select(tbl, f(A, B))
```

Users would need to be made aware that if they want 3VL semantics then they need to make all logical operators visible to the `@query` macro **OR** override the standard lifting behavior by defining `lift(f::type(f), xs...)`.

**Comment by nalimilan** [comment]
The strategy sounds good for expressions accessible to `@query`, but indeed the inconsistency with functions is a bit annoying. Of course I don't have a solution to that, except that we should probably make it easy (via a macro) to change the lifting behaviour of a function, either at the definition or at the call site: then the function would have to deal with `Nullable` itself.

**Comment by davidagold** [comment]
I agree that the inconsistency is annoying, but, as you suggest, we can remedy that with a macro (which, for this scheme, would really have to be at the definition site, and would have the same limitations there). 

However, as I try to implement this solution, I realize that the plot thickens. In order for `lift(f, xs...)` to be type-stable, we need to know the return type of `f` over the "eltype" argument signature of the `xs`, (i.e. `Core.Inference.return_type(f, map(x->eltype(typeof(x)), xs))`). However, these return types are invariant over the rows, hence we don't want to compute them for each row, i.e. inside of the `lift` function. Really, we want `lift(f, T, xs...)` (there's another discussion about how best to deal with the cost of splatting, but let's leave that aside for now), where we can compute `T` once and pass it to the respective `lift` invocation.

However, if we replace each function call in a row kernel expression -- e.g. `f(A) & g(b)` in `filter(tbl, f(A) & g(B))` -- with its lifted variant, then we need to compute the return type for each `f(A)`, `g(b)`, and `&(f(A), g(B))` and pass these computed return types to each respective `lift` invocation. Currently, the row kernel above would become, after processing,

``` julia
tuple -> &(f(tuple[1]), g(tuple[2]))
```

But in the proposed scheme, in order to accommodate return type arguments, which can only be calculated at run-time, we'd want the processed kernel to look something like

``` julia
(args_tuple, rt_tuple) -> 
    lift(
        &,
        rt_tuple[1],
        lift(f, rt_tuple[2], args_tuple[1]),
        lift(g, rt_tuple[3], args_tuple[2])
    )
```

where `args_tuple` is the tuple of arguments returned from the row iterator and `rt_tuple` is a tuple of return types calculated once before iterating over the rows.

cc @johnmyleswhite
